### PR TITLE
SkipScan

### PR DESCRIPTION
**Pull request by svenklemm**
This patch implements a skip-scan; an optimization for `SELECT DISTINCT ON`.
Usually for `SELECT DISTINCT ON` postgres will plan either a `UNIQUE` over a
sorted path, or some form of aggregate. In either case, it needs to scan the
entire table, even in cases where there are only a few unique values.

A skip-scan optimizes this case when we have an ordered index. Instead of
scanning the entire table and deduplicating after, the scan remembers the last
value returned, and searches the index for the next value after that one. This
means that for a table with `k` keys, with `u` distinct values, a skip-scan runs
in time `u * log(k)` as opposed to scanning then deduplicating, which takes time
`k`. We can write the number of unique values `u` as of function of `k` by
dividing by the number of repeats `r` i.e. `u = k/r` this means that a skip-scan
will be faster if each key is repeated more than a logarithmic number of times,
i.e. if `r > log(k)` then `u * log(k) < k/log(k) * log(k) < k`.

Disable-check: commit-count

### CODE REVIEW

**Comment by k-rus** [comment]
@svenklemm Will you add this feature to the change log?

**Comment by svenklemm** [comment]
> @svenklemm Will you add this feature to the change log?

done

File: tsl/src/nodes/skip_scan/exec.c [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+		 * skip previously seen values, we'll change back the number of scan keys
+		 * the first time through update_skip_key
+		 */
+		skip_scan_state_remove_skip_qual(state);
+		skip_scan_state_beginscan(state);
+
+		/* ignore the first scan key, which is the qual we add to skip repeat
+		 * values, the other quals still need to be applied. We'll set this back
+		 * once we have the inital values, and our qual can be applied.
+		 */
```
Comment (JLockerman) [inline tsl/src/nodes/skip_scan/exec.c:306]:
```suggestion
		/* ignore the first skip scan key, which is the qual we add to skip repeat
		 * values, the other quals still need to be applied. We'll set this back
		 * once we have the inital values, and our qual can be applied.
		 */
```
This is no longer accurate as the skip scan key does not need to be the first key

File: tsl/src/nodes/skip_scan/exec.c [hunk 2]
```diff
@@ -0,0 +1,597 @@
...
+	SkipScanState *state = (SkipScanState *) node;
+	TupleTableSlot *result;
+
+	if (skip_scan_is_searching_for_first_val(state))
+	{
+		Assert(skip_scan_state_get_scandesc(state) == NULL);
+		/* first time through we ignore the inital scan keys, which are used to
+		 * skip previously seen values, we'll change back the number of scan keys
+		 * the first time through update_skip_key
+		 */
```
Comment (JLockerman) [inline tsl/src/nodes/skip_scan/exec.c:299]:
```suggestion
		/* first time through we ignore the skip scan keys, which are used to
		 * skip previously seen values, we'll change back the number of scan keys
		 * the first time through update_skip_key
		 */
```
This is no longer accurate as the skip scan key does not need to be the first key

**Review by JLockerman — COMMENTED** [review]
Some comments I forgot to update

**Review by gayyappan — COMMENTED** [review]
Have to make another pass. Initial comments.

File: tsl/src/nodes/skip_scan/planner.c [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+	skip_scan_path->cpath.path.pathtype = T_CustomScan;
+	skip_scan_path->cpath.path.pathkeys = index_path->path.pathkeys;
+	skip_scan_path->cpath.path.pathtarget = index_path->path.pathtarget;
+	skip_scan_path->cpath.path.param_info = index_path->path.param_info;
+	skip_scan_path->cpath.path.parent = index_path->path.parent;
+	skip_scan_path->cpath.custom_paths = list_make1(index_path);
+	skip_scan_path->cpath.methods = &skip_scan_path_methods;
+
+	skip_scan_path->cpath.path.startup_cost = index_path->path.startup_cost;
+	skip_scan_path->cpath.path.total_cost = ndistinct * log2(index_path->path.total_cost);
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/planner.c:293]: done

File: tsl/src/nodes/skip_scan/planner.c [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+static CustomPathMethods skip_scan_path_methods = {
+	.CustomName = "SkipScanPath",
+	.PlanCustomPath = skip_scan_plan_create,
+};
+
+static SkipScanPath *skip_scan_path_create(PlannerInfo *root, IndexPath *index_path,
+										   double ndistinct);
+
+/*
+ * XXX: document created pathes/path transformations
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/planner.c]: fixed

File: tsl/src/nodes/skip_scan/planner.c [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+static SkipScanPath *
+skip_scan_path_create(PlannerInfo *root, IndexPath *index_path, double ndistinct)
+{
+	if (index_path->indexinfo->sortopfamily == NULL)
+		return NULL; /* non-orderable index, skip these for now */
+
+	/* we do not support orderByKeys out of conservatism; we do not know what,
+	 * if any, work would be required to support them.
+	 */
+	if (index_path->indexorderbys != NIL)
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/planner.c:278]: no this is actually something different. this is for indexes that support ordering by operator like ordering points by distance. most indexam do not support this,

File: tsl/src/nodes/skip_scan/planner.c [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+{
+	EquivalenceClass *ec = pathkey->pk_eclass;
+	TargetEntry *tle;
+
+	/* since SkipScan is on top of an index the EquivalenceClass
+	 * should not be volatile */
+	Assert(!ec->ec_has_volatile);
+
+	ListCell *lc;
+	foreach (lc, tlist)
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/planner.c:374]: tlist is not a single value

File: tsl/src/nodes/skip_scan/planner.c [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+
+static Plan *
+skip_scan_plan_create(PlannerInfo *root, RelOptInfo *relopt, CustomPath *best_path, List *tlist,
+					  List *clauses, List *custom_plans)
+{
+	SkipScanPath *path = (SkipScanPath *) best_path;
+	CustomScan *skip_plan = makeNode(CustomScan);
+	IndexPath *index_path = path->index_path;
+
+	tlist = ts_build_path_tlist(root, (Path *) path);
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/planner.c:64]: This was necessary before but is actually no longer necessary since the way we build the path prevents targetlist modifications by postgres

**Review by erimatnor — COMMENTED** [review]
Didn't do a full review, just posting some quick nits.

**Review by mkindahl — COMMENTED** [review]
Adding a few comments, but am not done reviewing. One question though. From the comments in `nodeIndexscan.c` we see that `ScanKey` is generated for the following cases:

1. Simple operator with constant comparison value
2. Simple operator with non-constant value
3. RowCompareExpr
4. ScalarArrayOpExpr
5. NullTest

You're clearly handling the first and second, and there were some issues with scalar array op expressions, but have you handled row compare expressions as well, or can that be an issue? From what I understand, these are generated from the `IndexPath` and can be

File: tsl/test/sql/include/skip_scan_query.sql [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+CREATE INDEX skip_scan_idx_dev_name ON :TABLE(dev_name);
+CREATE INDEX skip_scan_idx_dev_nulls_last ON :TABLE(dev);
+CREATE INDEX skip_scan_idx_dev_nulls_first ON :TABLE(dev NULLS FIRST);
+CREATE INDEX skip_scan_idx_dev_time_idx ON :TABLE(dev, time);
+CREATE INDEX skip_scan_idx_time_dev_idx ON :TABLE(time,dev);
+
+-- basic DISTINCT queries
+:PREFIX SELECT DISTINCT dev, 'q1_1' FROM :TABLE ORDER BY dev;
+:PREFIX SELECT DISTINCT dev_name, 'q1_2' FROM :TABLE ORDER BY dev_name;
+:PREFIX SELECT DISTINCT dev, 'q1_3', NULL FROM :TABLE ORDER BY dev;
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply tsl/test/sql/include/skip_scan_query.sql:59]: ... or at least, is there a test that it does not accidentally cause problems for skip scan?

File: tsl/test/sql/.gitignore [hunk 9]
```diff
@@ -11,6 +11,7 @@
 /dist_hypertable-*.sql
 /dist_query-*.sql
 /move-*.sql
+/plan_skip_scan-*.sql
 /reorder-*.sql
 /remote_copy-*.sql
 /jit-*.sql
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/test/sql/.gitignore:17]: This is completely unrelated to this PR

File: tsl/src/nodes/skip_scan/README.md [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+or
+
+```SQL
+Unique
+   ->  Merge Append
+         ->  Custom Scan (SkipScan)
+                ->  Index Scan
+                    Index Cond: (column > [prev column]) ...
+         ...
+```
```
[the remark this answers is not in the document]
Comment (k-rus) [reply tsl/src/nodes/skip_scan/README.md:75]: IMHO it is self-explanatory, since the merge append merges several scans, which is noted with `...`, and uniqueness in one plan doesn't guarantee uniqueness of the merge result.

So I suggest to keep as it is.

File: tsl/src/nodes/skip_scan/planner.c [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+		if (attno == idxinfo->indexkeys[i])
+			return i;
+	}
+
+	elog(ERROR, "column not present in index: %d", attno);
+	pg_unreachable();
+}
+
+/*
+ * find_ec_member_for_tle
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/planner.c:428]: this is a modified version so i think its better to keep it local, the normal find_ec_member_for_tle skips chunks but we dont want them skipped here cause our skipscan sits on top of chunk

File: tsl/src/nodes/skip_scan/planner.c [hunk 12]
```diff
@@ -0,0 +1,409 @@
...
+
+	/* find the ordering operator we'll use to skip around each key column */
+	PathKey *first_pathkey = linitial(index_path->path.pathkeys);
+	TargetEntry *tle = get_tle_for_pathkey(idx_info->indextlist, first_pathkey);
+
+	/* SkipScan on expressions not supported */
+	if (!IsA(tle->expr, Var))
+		return NULL;
+
+	Var *var = castNode(Var, tle->expr);
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/planner.c:312]: done

File: tsl/src/nodes/skip_scan/planner.c [hunk 13]
```diff
@@ -0,0 +1,458 @@
...
+							   var->varattno /*varattno*/,
+							   column_type /*vartype*/,
+							   -1 /*vartypmod*/,
+							   column_collation /*varcollid*/,
+							   0 /*varlevelsup*/);
+
+	Expr *comparison_expr = make_opclause(comparator,
+										  BOOLOID /*opresulttype*/,
+										  false /*opretset*/,
+										  &current_val->xpr /*leftop*/,
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/planner.c:349]: saves us from having to cast

File: tsl/test/sql/include/skip_scan_query.sql (hunk 8 above)
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/test/sql/include/skip_scan_query.sql:59]: there are multiple tests without explicit ORDER BY

File: tsl/src/nodes/skip_scan/exec.c [hunk 14]
```diff
@@ -0,0 +1,595 @@
...
+	if (*state->num_scan_keys <= 0)
+		return;
+
+	state->skip_qual = (*state->scan_keys)[0];
+	state->skip_qual_offset = 0;
+
+	skip_scan_state_fixup_qual_order(state, runtime_keys, num_runtime_keys);
+}
+
+/* ScanKeys must be ordered by index attribute, while we put the skip qual at the
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/exec.c]: this code has been refactored and qual ordering happens in the planner bits now

File: tsl/test/sql/include/skip_scan_query.sql [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+
+-- ORDER BY
+:PREFIX SELECT time, dev, val, 'q4_1' FROM (SELECT DISTINCT ON (dev) * FROM :TABLE ORDER BY dev, time) a;
+:PREFIX SELECT time, dev, val, 'q4_2' FROM (SELECT DISTINCT ON (dev) * FROM :TABLE ORDER BY dev DESC, time DESC) a;
+
+-- WHERE CLAUSES
+:PREFIX SELECT time, dev, val, 'q5_1' FROM (SELECT DISTINCT ON (dev) * FROM :TABLE WHERE dev > 5) a;
+:PREFIX SELECT time, dev, val, 'q5_2' FROM (SELECT DISTINCT ON (dev) * FROM :TABLE WHERE time > 5) a;
+:PREFIX SELECT time, dev, val, 'q5_1' FROM (SELECT DISTINCT ON (dev) * FROM :TABLE) a WHERE dev > 5;
+:PREFIX SELECT time, dev, val, 'q5_2' FROM (SELECT DISTINCT ON (dev) * FROM :TABLE) a WHERE time > 5;
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/test/sql/include/skip_scan_query.sql]: done

File: tsl/src/nodes/skip_scan/exec.c [hunk 16]
```diff
@@ -0,0 +1,595 @@
...
+ * front so it's easy to find. Now that it's in an easy-to-work-with form, move
+ * the skip key if the distinct-column is not the first one in the index
+ */
+static void
+skip_scan_state_fixup_qual_order(SkipScanState *state, IndexRuntimeKeyInfo *runtime_keys,
+								 int num_runtime_keys)
+{
+	/* find the correct location for the skip qual, it should be the first qual
+	 * on its column
+	 */
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/exec.c]: done

File: tsl/src/nodes/skip_scan/planner.c [hunk 17]
```diff
@@ -0,0 +1,458 @@
...
+	skip_scan_path->cpath.path.pathtype = T_CustomScan;
+	skip_scan_path->cpath.path.pathkeys = index_path->path.pathkeys;
+	skip_scan_path->cpath.path.pathtarget = index_path->path.pathtarget;
+	skip_scan_path->cpath.path.param_info = index_path->path.param_info;
+	skip_scan_path->cpath.path.parent = index_path->path.parent;
+	skip_scan_path->cpath.custom_paths = list_make1(index_path);
+	skip_scan_path->cpath.methods = &skip_scan_path_methods;
+
+	Assert(index_path->path.total_cost > 0);
+	Selectivity s1 = clauselist_selectivity(root,
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/planner.c:288]: done

File: tsl/test/sql/include/skip_scan_load.sql [hunk 18]
```diff
@@ -0,0 +1,33 @@
+-- This file and its contents are licensed under the Timescale License.
+-- Please see the included NOTICE for copyright information and
+-- LICENSE-TIMESCALE for a copy of the license.
+
+CREATE TABLE skip_scan(time int, dev int, dev_name text, val int);
+
+INSERT INTO skip_scan SELECT t, d, 'device_' || d::text, random() FROM generate_series(1, 1000) t, generate_series(1, 10) d;
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/test/sql/include/skip_scan_load.sql]: What exactly are you suggesting here?

File: tsl/test/sql/include/skip_scan_query.sql [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+:PREFIX SELECT DISTINCT dev FROM :TABLE ORDER BY dev;
+:PREFIX SELECT DISTINCT dev FROM :TABLE ORDER BY dev DESC;
+:PREFIX SELECT DISTINCT ON (dev) dev FROM :TABLE;
+DROP INDEX skip_scan_idx_dev_nulls_last;
+
+-- NULLS FIRST index on dev
+CREATE INDEX skip_scan_idx_dev_nulls_first ON :TABLE(dev NULLS FIRST);
+:PREFIX SELECT DISTINCT dev FROM :TABLE ORDER BY dev NULLS FIRST;
+:PREFIX SELECT DISTINCT ON (dev) dev FROM :TABLE ORDER BY dev NULLS FIRST;
+DROP INDEX skip_scan_idx_dev_nulls_first;
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/test/sql/include/skip_scan_query.sql:23]: there are tests for NULL-handling

File: tsl/src/nodes/skip_scan/exec.c [hunk 20]
```diff
@@ -0,0 +1,316 @@
...
+				 * if we found a tuple we update the skip scan key
+				 * and look for values greater than the value we just
+				 * found. If this is the first non-NULL value we
+				 * also switch stage to look for values greater than
+				 * that in subsequent calls.
+				 */
+				if (state->stage == SS_NOT_NULL)
+					skip_scan_switch_stage(state, SS_VALUES);
+
+				skip_scan_update_key(state, result);
```
Comment (gayyappan) [inline tsl/src/nodes/skip_scan/exec.c:230]: return result should be right after 230, not after else? If we don't have  a valid tuple, we should not return at 251 and just switch states.

File: tsl/src/nodes/skip_scan/exec.c [hunk 21]
```diff
@@ -0,0 +1,316 @@
...
+
+	state->stage = new_stage;
+}
+
+static void
+skip_scan_update_key(SkipScanState *state, TupleTableSlot *slot)
+{
+	if (!state->prev_is_null && !state->distinct_by_val)
+		pfree(DatumGetPointer(state->prev_datum));
+
```
Comment (gayyappan) [inline tsl/src/nodes/skip_scan/exec.c:169]: We get here only if state is SS_VALUES. Should add an assert for it.

File: tsl/src/nodes/skip_scan/exec.c [hunk 22]
```diff
@@ -0,0 +1,316 @@
...
+	MemoryContext old_ctx = MemoryContextSwitchTo(state->ctx);
+	state->prev_datum = slot_getattr(slot, state->distinct_col_attnum, &state->prev_is_null);
+	if (state->prev_is_null)
+	{
+		state->skip_key->sk_flags = SK_ISNULL;
```
Comment (gayyappan) [inline tsl/src/nodes/skip_scan/exec.c:174]: The skip_key->sk_flags should only be overwritten when we switch states (and already handled there) and this code sets the sk_argument.  I don't understand why skip_key->sk_flags is set to SK_ISNULL here. Why is this needed?

File: tsl/src/nodes/skip_scan/exec.c [hunk 23]
```diff
@@ -0,0 +1,316 @@
...
+		default:
+			return NULL;
+			break;
+	}
+}
+
+static void
+skip_scan_end(CustomScanState *node)
+{
+	SkipScanState *state = (SkipScanState *) node;
```
Comment (gayyappan) [inline tsl/src/nodes/skip_scan/exec.c:269]: Assert that state is SS_END here.

**Review by gayyappan — COMMENTED** [review]
The executor rewrite is so much easier to follow!

File: tsl/src/nodes/skip_scan/exec.c (hunk 23 above)
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/exec.c:269]: that would not be correct, SS_END means the scan got exhausted but the scan can be ended before that by parent node e.g. a LIMIT node that has reached the limit

File: tsl/src/nodes/skip_scan/exec.c (hunk 20 above)
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/exec.c:230]: we need to return there even for SS_END but i can put return on line 231 and after 249 to make it easier to follow

Comment (svenklemm) [reply tsl/src/nodes/skip_scan/exec.c:230]: or actually what you suggest would work too with another round trip to skip_scan_exec
current code saves us 1 roundtrip for the SS_END case

File: tsl/src/nodes/skip_scan/exec.c (hunk 22 above)
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/exec.c:174]: SK_ISNULL is about the state of sk_argument, not setting SK_ISNULL here can lead to segfault for pointer datum types. but it's all theoretical since we should never arrive in that branch since the skip qual will prevent use from getting NULL values and we scan for NULL exlicitly

File: tsl/test/sql/include/skip_scan_query_ht.sql [hunk 24]
```diff
@@ -0,0 +1,24 @@
+-- This file and its contents are licensed under the Timescale License.
+-- Please see the included NOTICE for copyright information and
+-- LICENSE-TIMESCALE for a copy of the license.
+
+CREATE INDEX ON :TABLE(dev);
+CREATE INDEX ON :TABLE(time);
```
Comment (gayyappan) [inline tsl/test/sql/include/skip_scan_query_ht.sql:6]: We should also add a test for skipscan with index on character datatypes.

**Comment by mfundul** [comment]
There are some usages of the letters "pathes" in the comment, e.g. "IndexPathes". Are those typos?

File: tsl/src/nodes/skip_scan/planner.c [hunk 25]
```diff
@@ -0,0 +1,599 @@
...
+		{
+			SkipScanPath *skip_path =
+				skip_scan_path_create(root, castNode(IndexPath, child), ndistinct);
+
+			if (skip_path)
+			{
+				child = (Path *) skip_path;
+				has_skip_path = true;
+			}
+		}
```
Comment (mkindahl) [inline tsl/src/nodes/skip_scan/planner.c:400]: Why does this only check for `IndexPath` while the "upper" function has checks for `IndexOnlyPath` as well?

File: tsl/src/nodes/skip_scan/planner.c [hunk 26]
```diff
@@ -0,0 +1,599 @@
...
+	{
+		if (indexclauses[i] != NIL)
+			ordered_list = list_concat(ordered_list, indexclauses[i]);
+	}
+
+	return ordered_list;
+}
+
+static OpExpr *
+fiX_indexqual(IndexOptInfo *index, RestrictInfo *rinfo, AttrNumber distinct_column)
```
Comment (mkindahl) [inline tsl/src/nodes/skip_scan/planner.c:573]: Typo?
```suggestion
fix_indexqual(IndexOptInfo *index, RestrictInfo *rinfo, AttrNumber distinct_column)
```

**Review by mkindahl — COMMENTED** [review]
I think the stage-switching function is hard to follow because of the interleaved conditions and recursion. It would be great it it could be made easier to follow. It does look correct though, but I'm concerned about how easy it is to maintain. The other things are minor things.

File: tsl/src/nodes/skip_scan/exec.c [hunk 27]
```diff
@@ -0,0 +1,316 @@
...
+		state->scan_keys = &idx->ioss_ScanKeys;
+		state->num_scan_keys = &idx->ioss_NumScanKeys;
+		state->scan_desc = &idx->ioss_ScanDesc;
+	}
+	else
+		elog(ERROR, "unknown subscan type in SkipScan");
+
+	/* scankeys are not setup for explain only */
+	if (eflags & EXEC_FLAG_EXPLAIN_ONLY)
+		return;
```
[the remark this answers is not in the document]
Comment (davidkohn88) [reply tsl/src/nodes/skip_scan/exec.c:88]: Does this mean that SkipScans won't show up in explains (but they will in explain analyze?)

[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/exec.c:88]: no it means we skip initializing some internal stuff if its explain only but it doesnt affect what is shown in EXPLAIN

[the remark this answers is not in the document]
Comment (davidkohn88) [reply tsl/src/nodes/skip_scan/exec.c:88]: 👍

File: tsl/src/nodes/skip_scan/planner.c [hunk 28]
```diff
@@ -0,0 +1,595 @@
...
+ * SkipScan Plan Creation *
+ **************************/
+
+static CustomScanMethods skip_scan_plan_methods = {
+	.CustomName = "SkipScan",
+	.CreateCustomScanState = tsl_skip_scan_state_create,
+};
+
+void
+_skip_scan_init(void)
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/planner.c:56]: this is done in all init functions

File: tsl/src/nodes/skip_scan/planner.c (hunk 25 above)
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/planner.c:400]: there is no IndexOnlyPath

File: tsl/src/nodes/skip_scan/planner.c (hunk 28 above)
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/planner.c:56]: check src/init.c for more examples of this

File: tsl/test/sql/include/skip_scan_query_ht.sql (hunk 24 above)
Comment (svenklemm) [reply tsl/test/sql/include/skip_scan_query_ht.sql:6]: there are tests for text column (dev_name)

File: tsl/src/nodes/skip_scan/exec.c [hunk 29]
```diff
@@ -0,0 +1,316 @@
...
+			/*
+			 * if we found a NULL value we return it, otherwise
+			 * we restart the scan looking for non-NULL
+			 */
+			skip_scan_switch_stage(state, SS_NOT_NULL);
+			skip_scan_rescan_index(state);
+			if (!TupIsNull(result))
+				return result;
+			else
+				return skip_scan_exec(node);
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/exec.c:210]: while (TupIsNull(result)) would not work and it would have to be while(true) so you can pick between while(true) and recursion

File: tsl/src/nodes/skip_scan/exec.c [hunk 30]
```diff
@@ -0,0 +1,316 @@
...
+
+	/* find position of our skip key
+	 * skip key is put as first key for the respective column in sort_indexquals
+	 */
+	ScanKey data = *state->scan_keys;
+	for (int i = 0; i < *state->num_scan_keys; i++)
+	{
+		if (data[i].sk_flags == SK_ISNULL && data[i].sk_attno == state->sk_attno)
+		{
+			state->skip_key = &data[i];
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/exec.c:98]: newNode in tsl_skip_scan_state_create does palloc0

**Review by gayyappan — APPROVED** [review]
LGTM. Question about call to rescan_index.

File: tsl/src/nodes/skip_scan/exec.c [hunk 31]
```diff
@@ -0,0 +1,316 @@
...
+skip_scan_rescan(CustomScanState *node)
+{
+	SkipScanState *state = (SkipScanState *) node;
+	/* reset stage so we can assert in skip_scan_switch_stage that stage always moves forward */
+	state->stage = SS_BEGIN;
+
+	if (has_nulls_first(state))
+		skip_scan_switch_stage(state, SS_NULLS_FIRST);
+	else
+		skip_scan_switch_stage(state, SS_NOT_NULL);
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/exec.c:289]: i've added a comment why we do this

File: tsl/src/nodes/skip_scan/planner.c (hunk 25 above)
Comment (mkindahl) [reply tsl/src/nodes/skip_scan/planner.c:400]: Ah, mixed them up. Was thinking of `IndexOnlyScan`, which is not relevant here.

File: tsl/test/sql/include/skip_scan_query.sql (hunk 19 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply tsl/test/sql/include/skip_scan_query.sql:23]: The state machine has checks for tables that starts with NULLs and ends with NULLs (but contain other non-null values for the column that is being scanned), but I see no test with a table that has *only* NULLs, which could be a potentially bad case for the state machine.

**Review by mkindahl — APPROVED** [review]
I do not see a check for a table with *only* null values in it, which could be a potentially bad case for the state machine. You have tests for tables that contain a mix of null and non-null values, but I do not see a test where the column has only nulls.

I'm approving anyway, since this is a minor thing, but please add a test for that.

File: tsl/src/nodes/skip_scan/exec.c [hunk 32]
```diff
@@ -0,0 +1,316 @@
...
+			else
+				return skip_scan_exec(node);
+
+			break;
+
+		case SS_NOT_NULL:
+		case SS_VALUES:
+			result = state->idx->ps.ExecProcNode(&state->idx->ps);
+
+			if (!TupIsNull(result))
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/nodes/skip_scan/exec.c]: i introduced needs_rescan flag to switch this around
